File: Array/Python3/H41-first-missing-positive.py
# time:  O(n)
# space: O(1)

class Solution:
    def firstMissingPositive(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        nums.append(0)
        l = len(nums)
        if l == 1:
            return 1

        for i in range(l):
            if nums[i] < 0 or nums[i] >= l:
                nums[i] = 0
                
        for i in range(l):
            nums[nums[i] % l] += l
            
        for i in range(1, l):
            if nums[i] // l == 0:
                return i
        return l

File: HashTable/C/H41-first-missing-positive.c
// time:  O(n)
// space: O(1)

int firstMissingPositive(int* nums, int numsSize) {
    if(numsSize == 0) return 1;
    int numsi = 0;
    numsSize++;
    for(int i = 0; i < numsSize-1; i++)
        if(nums[i] < 0 || nums[i] >= numsSize)
            nums[i] = 0;
    
    for(int i = 0; i < numsSize-1; i++)
        if(nums[i] % numsSize < numsSize-1)
            nums[nums[i] % numsSize] += numsSize;
        else numsi += numsSize;
        
    for(int i = 1; i < numsSize-1; i++)
        if(nums[i] / numsSize == 0)
            return i;
    if(numsi / numsSize == 0)
        return numsSize - 1;
    return numsSize;
}